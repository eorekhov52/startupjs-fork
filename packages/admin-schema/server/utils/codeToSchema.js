import { parseSync } from '@babel/core'
import _traverseModule from '@babel/traverse'
import { SCHEMA_EXPORT_NAME } from '../constants.js'
import { ASSOCIATION_SPREAD_KEY, ASSOCIATION_TYPES } from '../../isomorphic/constants.js'

// traverse might not be available as an ESM default import
const traverse = typeof _traverseModule === 'function' ? _traverseModule : _traverseModule.default

export default function codeToSchema (code) {
  const ast = parseSync(code, {
    sourceType: 'module',
    babelrc: false,
    configFile: false
  })
  let schema
  traverse(ast, {
    // find a named export (export const schema = { ... })
    ExportNamedDeclaration ($this) {
      const $declaration = $this.get('declaration')
      if (!$declaration.node) return
      if (!$declaration.isVariableDeclaration()) return
      const $declarator = $declaration.get('declarations.0')
      if (!$declarator.node) return
      if ($declarator.get('id').node.name !== SCHEMA_EXPORT_NAME) return
      const $init = $declarator.get('init')
      if (!$init.node) return
      if (!$init.isObjectExpression()) return
      schema = parseValue($init)
    }
  })
  return schema
}

// recursively parse object expression into JSON object
function parseValue ($value) {
  if ($value.isObjectExpression()) {
    const object = {}
    for (const $property of $value.get('properties')) {
      if ($property.isSpreadElement()) {
        const [key, value] = parseAssociation($property)
        object[key] = value
        continue
      }
      if (!$property.isObjectProperty()) throw Error('Unsupported property type:\n' + $property.toString())
      const $key = $property.get('key')
      let key
      if ($key.isIdentifier()) {
        key = $key.node.name
      } else if ($key.isLiteral()) {
        key = $key.node.value
      } else {
        throw Error('Unsupported key type:\n' + $key.toString())
      }
      const value = parseValue($property.get('value'))
      object[key] = value
    }
    return object
  } else if ($value.isLiteral()) {
    return $value.node.value
  } else if ($value.isArrayExpression()) {
    return $value.get('elements').map($element => parseValue($element))
  } else {
    throw Error('Unsupported value type:\n' + $value.toString())
  }
}

// spread element is used to extend field with the association properties
// e.g. { ...belongsTo('events'), required: true }
//
// Transforms:
//   ...belongsTo('events')
// to:
//   '$$code_spread_association', { type: 'belongsTo', collection: 'events' }
function parseAssociation ($spreadElement) {
  const $argument = $spreadElement.get('argument')
  if (!$argument.isCallExpression()) {
    throw Error('Association must be done through spreading the function. Got:\n' + $argument.toString())
  }
  const $callee = $argument.get('callee')
  if (!$callee.isIdentifier()) {
    throw Error('Association function must be an imported identifier. Got:\n' + $callee.toString())
  }
  const type = $callee.node.name
  if (!ASSOCIATION_TYPES.includes(type)) {
    throw Error(`Unsupported association type.\nSupported ${ASSOCIATION_TYPES}.\nGot:\n` + type)
  }
  const $arguments = $argument.get('arguments')
  if ($arguments.length !== 1) {
    throw Error('Association function accepts collection name as a single argument. Got:\n' + $arguments.toString())
  }
  const $collection = $arguments[0]
  if (!$collection.isStringLiteral()) {
    throw Error('Association collection must be a string. Got:\n' + $arguments[0].toString())
  }
  const collection = $collection.node.value
  return [ASSOCIATION_SPREAD_KEY, { type, collection }]
}
