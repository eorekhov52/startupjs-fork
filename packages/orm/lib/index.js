import { singletonMemoize } from '@startupjs/cache'
import racer from 'racer'
import promisifyRacer from './promisifyRacer.js'

const Model = racer.Model

global.STARTUP_JS_ORM = {}

export default function (racer) {
  if (racer.orm) return

  racer._orm = global.STARTUP_JS_ORM
  racer.orm = function (pattern, OrmEntity, alias) {
    const name = alias || pattern
    if (global.STARTUP_JS_ORM[name]) throw alreadyDefinedError(pattern, alias)

    // NOTE
    // if same OrmEntity will be passed for different collections
    // then they will all have the same collection name
    if (!OrmEntity.collection) {
      const match = pattern.match(/^[^.]+/)
      if (match) OrmEntity.collection = match[0]
    }

    global.STARTUP_JS_ORM[name] = {
      pattern,
      regexp: patternToRegExp(pattern),
      OrmEntity
    }
  }

  Model.prototype.at = function (subpath, alias) {
    const path = this.path(subpath)
    return this.scope(path, alias)
  }

  Model.prototype._scope = function (path) {
    const ChildModel = Model.ChildModel
    const model = new ChildModel(this)
    model._at = path
    return model
  }

  // TODO: do NOT cache within built-in use* hooks from react-sharedb
  //       because they already handle caching internally
  Model.prototype.scope = singletonMemoize(function (path, alias) {
    if (alias) {
      if (global.STARTUP_JS_ORM[alias]) {
        return this.__createScopedModel(path, global.STARTUP_JS_ORM[alias].OrmEntity)
      } else {
        throw new Error(
          'Non-existent alias of the OrmEntity specified: ' +
            alias +
            '\n\n' +
            'Most likely you have specified the path incorrectly in ' +
            '".scope()" or ".at()"\n\n' +
            'The path must be passed as a single string separated by dots, ' +
            'for example:\n\n' +
            'CORRECT:\n' +
            "$root.at('users.' + userId)\n\n" +
            'INCORRECT:\n' +
            "$root.at('users', userId)"
        )
      }
    }

    const segments = this._dereference(this.__splitPath(path), true)
    const fullPath = segments.join('.')

    for (const name in global.STARTUP_JS_ORM) {
      const regexp = global.STARTUP_JS_ORM[name].regexp
      if (regexp.test(fullPath)) {
        return this.__createScopedModel(path, global.STARTUP_JS_ORM[name].OrmEntity)
      }
    }

    return this._scope(path)
  }, {
    cacheName: 'model',
    nestedThis: true
  })

  Model.prototype.__createScopedModel = function (path, OrmEntity) {
    let model
    if (OrmEntity.factory || OrmEntity.prototype.factory) {
      model = OrmEntity(this._scope(path), this)
      // if factory didn't return anything, return a simple scoped model
      if (!model) model = this._scope(path)
    } else {
      model = new OrmEntity(this)
    }
    model._at = path
    return model
  }

  Model.prototype.__splitPath = function (path) {
    return (path && path.split('.')) || []
  }

  promisifyRacer()
}

function patternToRegExp (pattern) {
  pattern = pattern.replace(/\$/g, '\\$').replace(/\./g, '\\.').replace(/\*/g, '([^\\.]*)')
  return new RegExp('^' + pattern + '$')
}

function alreadyDefinedError (pattern, alias) {
  let msg
  if (alias) {
    msg =
      "ORM entity with the alias '" +
      alias +
      "' is already defined. " +
      'Aliases must be unique. If you did already define the same ORM entity with ' +
      "that alias name, just don't specify the alias at all -- path pattern is sufficient."
  } else {
    msg = "ORM entity matching the same path pattern '" + pattern + "' is already defined."
  }
  return new Error(msg)
}

export const ChildModel = Model.ChildModel

function BaseModel () {
  Model.ChildModel.apply(this, arguments)
}

BaseModel.prototype = Object.create(Model.ChildModel.prototype)
BaseModel.prototype.constructor = BaseModel

BaseModel.prototype.getId = function () {
  const actualField = this.dereferenceSelf()
  return actualField.leaf()
}

BaseModel.prototype.getCollection = function () {
  let collection = this.constructor.collection

  // fallback when orm is factory
  if (!collection) {
    const model = this.root
    const actualField = this.dereferenceSelf()
    collection = model._splitPath(actualField.path())[0]
  }

  return collection
}

BaseModel.prototype.dereferenceSelf = function () {
  const model = this.root
  const segments = model._splitPath(this.path())
  return model.scope(model._dereference(segments, true).join('.'))
}

BaseModel.associations = []

BaseModel.addAssociation = function (association) {
  this.associations = this.associations.concat(association)
}

BaseModel.prototype.getAssociations = function () {
  return this.constructor.associations
}

export { BaseModel }
export * from './associations/index.js'

/**
 * Pick properties from json-schema to be used in a form.
 * Supports simplified schema (just the properties object) and full schema.
 * Performs extra transformations like auto-generating `label`.
 * `createdAt`, `updatedAt`, `id`, `_id` fields are excluded by default.
 * @param {Object} schema
 * @param {Object|Array} options - exclude or include fields. If array, it's the same as passing { include: [...] }
 * @param {Array} options.include - list of fields to pick (default: all)
 * @param {Array} options.exclude - list of fields to exclude (default: none)
 */
export function pickFormFields (schema, options) {
  let include, exclude
  if (Array.isArray(options)) {
    include = options
  } else {
    ;({ include, exclude } = options || {})
  }
  exclude ??= []
  if (!schema) throw Error('pickFormFields: schema is required')
  schema = JSON.parse(JSON.stringify(schema))
  if (schema.type === 'object') {
    schema = schema.properties
  }
  for (const key in schema) {
    if (
      include?.includes(key) ||
      (!include && !exclude.includes(key) && !DEFAULT_EXCLUDE_FORM_FIELDS.includes(key))
    ) {
      const field = schema[key]
      if (!field.label) field.label = camelCaseToLabel(key)
    } else {
      delete schema[key]
    }
  }
  return schema
}

const DEFAULT_EXCLUDE_FORM_FIELDS = ['id', '_id', 'createdAt', 'updatedAt']

// split into words, capitalize first word, make others lowercase
function camelCaseToLabel (str) {
  return str
    .replace(/([A-Z])/g, ' $1')
    .toLowerCase()
    .replace(/^./, (s) => s.toUpperCase())
}
