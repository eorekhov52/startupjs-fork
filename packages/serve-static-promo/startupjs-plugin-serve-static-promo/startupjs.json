{
  "type": "plugin",
  "name": "startupjs-plugin-serve-static-promo",
  "for": "startupjs"
}
