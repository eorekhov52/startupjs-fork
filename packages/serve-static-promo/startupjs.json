{
  "type": "module",
  "name": "serve-static-promo",
  "bundler": {
    "forceCompile": {
      "web": true,
      "server": true
    }
  },
  "imports": [
    "./startupjs-plugin-serve-static-promo"
  ]
}
