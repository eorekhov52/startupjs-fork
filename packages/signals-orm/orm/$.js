// this is just the $() function implementation.
// The actual $ exported from this package is a Proxy targeting the dataTree root,
// and this function is an implementation of the `apply` handler for that Proxy.
import { observe, unobserve } from '@nx-js/observer-util'
import getSignal from './getSignal.js'
import { set as _set, del as _del } from './dataTree.js'

export const LOCAL = '$local'
const DELETION_DELAY = 0
let counter = 0

const valuesFr = new FinalizationRegistry(id => {
  _del([LOCAL, id])
})

const reactionsFr = new FinalizationRegistry(([reaction, id]) => {
  unobserve(reaction)
  // don't delete data right away to prevent dependent reactions which are also going to be GC'ed
  // from triggering unnecessarily
  setTimeout(() => _del([LOCAL, id]), DELETION_DELAY)
})

export default function $ (value) {
  if (typeof value === 'function') {
    return reaction$(value)
  } else {
    return value$(value)
  }
}

function value$ (value) {
  const id = `_${counter}`
  counter += 1
  _set([LOCAL, id], value)
  const $value = getSignal([LOCAL, id])
  valuesFr.register($value, id)
  return $value
}

function reaction$ (fn) {
  const id = `_${counter}`
  counter += 1
  const reactionScheduler = reaction => runReaction(reaction, id)
  const reaction = observe(fn, { lazy: true, scheduler: reactionScheduler })
  runReaction(reaction, id)
  const $value = getSignal([LOCAL, id])
  reactionsFr.register($value, [reaction, id])
  return $value
}

function runReaction (reaction, id) {
  const newValue = reaction()
  _set([LOCAL, id], newValue)
}
