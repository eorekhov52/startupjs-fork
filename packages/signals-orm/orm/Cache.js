export default class Cache {
  constructor () {
    this.cache = new Map()
    this.fr = new FinalizationRegistry((...args) => this.delete(...args))
  }

  get (key) {
    return this.cache.get(key)?.deref()
  }

  set (key, value) {
    this.cache.set(key, new WeakRef(value))
    this.fr.register(value, key)
  }

  delete (key) {
    this.cache.delete(key)
  }

  get size () {
    return this.cache.size
  }
}
