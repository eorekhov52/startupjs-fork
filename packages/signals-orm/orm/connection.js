export let connection

export function setConnection (_connection) {
  connection = _connection
}

export function getConnection () {
  if (!connection) throw Error(ERRORS.notSet)
  return connection
}

const ERRORS = {
  notSet: `
    Connection is not set.
    You must set the initialized ShareDB connection before using subscriptions.
    You've probably forgotten to call connect() in your app:

    import connect from '@startupjs/signals-orm/connect'
    connect({ baseUrl: 'http://localhost:3000' })
  `
}
