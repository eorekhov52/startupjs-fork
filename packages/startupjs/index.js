// Make ShareDB client and Racer client work in React Native and in Webpack
import dummyMockBrowserify from '@startupjs/utils/mockBrowserify'
import dummyInit from '@startupjs/init/client-auto'

export { default as axios } from '@startupjs/utils/axios'
export * from '@startupjs/react-sharedb'
export { default as $, signal } from '@startupjs/signals'
export * from '@startupjs/hooks'
// HINT: `isomorphic` means that the code can be executed both
//        on the server and on the client
export * from '@startupjs/isomorphic-helpers'
// dummy babel macro functions for @startupjs/babel-plugin-rn-stylename-inline.
export function css (cssString) { return cssString }
export function styl (stylString) { return stylString }
// dummy pug function (it gets compiled to jsx by babel pug plugin)
export function pug (pugString) { return pugString }
export { default as t } from '@startupjs/i18n/client/t.js'

// this library is pure side-effects, so we force usage of a dummy function
// to prevent dead-code elimination
dummyMockBrowserify()
dummyInit() // prevent dead-code elimination
