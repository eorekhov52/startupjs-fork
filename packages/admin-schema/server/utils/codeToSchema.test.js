/* global it, expect */

import codeToSchema from './codeToSchema.js'

it('schema as a plain JSON-like object', () => {
  const code = /* js */`
    export const schema = {
      name: { type: 'string' },
      age: { type: 'number', min: 0, max: 100 },
      isStudent: { type: 'boolean' },
      tags: { type: 'array', items: { type: 'string' } }
    }
  `
  expect(codeToSchema(code)).toMatchSnapshot()
})

it('with spreads on associations (belongsTo, hasOne, hasMany)', () => {
  const code = /* js */`
    import { belongsTo, hasOne, hasMany, hasManyFlags } from 'startupjs/orm'
    export const schema = {
      eventId: {
        ...belongsTo('events'),
        required: true
      },
      "name": { type: 'string', required: true },
      "photoFileId": {
        ...hasOne('files'),
        input: 'file',
        label: 'Photo',
        mimeTypes: 'image/*'
      },
      likes: {
        ...hasMany('users'),
        input: 'likes'
      },
      likesFlags: {
        ...hasManyFlags('persons'),
        $comment: 'true flags for everyone this person likes',
        input: 'likesFlags'
      },
    }
  `
  expect(codeToSchema(code)).toMatchSnapshot()
})
