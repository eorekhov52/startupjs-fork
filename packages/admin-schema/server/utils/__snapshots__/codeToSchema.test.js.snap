// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema as a plain JSON-like object 1`] = `
{
  "age": {
    "max": 100,
    "min": 0,
    "type": "number",
  },
  "isStudent": {
    "type": "boolean",
  },
  "name": {
    "type": "string",
  },
  "tags": {
    "items": {
      "type": "string",
    },
    "type": "array",
  },
}
`;

exports[`with spreads on associations (belongsTo, hasOne, hasMany) 1`] = `
{
  "eventId": {
    "$$code_spread_association": {
      "collection": "events",
      "type": "belongsTo",
    },
    "required": true,
  },
  "likes": {
    "$$code_spread_association": {
      "collection": "users",
      "type": "hasMany",
    },
    "input": "likes",
  },
  "likesFlags": {
    "$$code_spread_association": {
      "collection": "persons",
      "type": "hasManyFlags",
    },
    "$comment": "true flags for everyone this person likes",
    "input": "likesFlags",
  },
  "name": {
    "required": true,
    "type": "string",
  },
  "photoFileId": {
    "$$code_spread_association": {
      "collection": "files",
      "type": "hasOne",
    },
    "input": "file",
    "label": "Photo",
    "mimeTypes": "image/*",
  },
}
`;
